server/src/main/java/org/opensearch/common/time/DateFormatter.java

```diff
@@ -164,7 +146,12 @@ static DateFormatter forPattern(String input, String printPattern, Boolean canCa
         // support the 6.x BWC compatible way of parsing java 8 dates
         String format = strip8Prefix(input);
         List<String> patterns = splitCombinedPatterns(format);
-        List<DateFormatter> formatters = patterns.stream().map(DateFormatters::forPattern).collect(Collectors.toList());
+        List<DateFormatter> formatters = patterns.stream().map(p -> {
```

Comment:
Super minor, I think we don't need to evaluate the version for each pattern but do this only once:

```
        if (supportedVersion.before(Version.V_2_12_0)) {
            patterns = patterns
                .stream()
                .map(LegacyFormatNames::camelCaseToSnakeCase)
                .collect(Collectors.toList());
        }
```

Reply:
As I as side note (curious to hear your thoughts), it looks to me the version quite pollutes the `DateFormatter` API, I was thinking what if we make it explicit this way:

```
interface VersionedDateFormatter {
    DateFormatter forVersion(final Version supportedVersion);
}
```

It is a little more work but there is good separation between DateFormatter and Version

Reply:
> ...it looks to me the version quite pollutes the DateFormatter API

I don't think it pollutes the API. The `forPattern` method is package private and only called from the private `DateFieldMapper.buildFormatter` method. It's not public so it shouldn't be an API concern. We can also address in a follow up issue if needed.

Reply:
> I don't think it pollutes the API. The `forPattern` method is package private

It is not package private sadly, it is defined in the interface (`DateFormatter` is an interface) and is public by default.

Reply:
Oh that's right it's an unsealed interface. I added back the unused method. There are a lot of unused methods like this in the formatter, and it's not intended to be a public facing API so it will need cleaning up in a followup PR anyway. We can look at formalizing an API then if we really think it's needed.